### PR TITLE
feat(compiler): Add `@unsafe` attribute for low-level code

### PR DESCRIPTION
**Pull request by jozanza**
This PR separates how Grain tracks unsafe wasm values (`WasmI32`, etc.) from how Grain tracks proper Grain values. This makes it far easier to write low-level code in Grain, as the GC won't attempt to collect random values that sort of look like pointers. Of course, Grain is not meant for low-level programming, so the non-memory related limitations still remain (unsafe values cannot be printed or used with generic functions). For this reason, the `@unsafe` attribute is required on functions that use unsafe values. 

It should be noted that this is a change to Grain as a whole, and `@unsafe` doesn't have any effect on how the function is compiled; it only suppresses the warnings emitted for using unsafe values. We may remove `@disableGC` at some point once all Grain runtime/stdlib code has been moved over to `@unsafe`.

### CODE REVIEW

File: compiler/src/typed/typed_well_formedness.re [hunk 1]
```diff
@@ -154,7 +154,8 @@ module WellFormednessArg: TypedtreeIter.IteratorArgument = {
 module WellFormednessIterator = TypedtreeIter.MakeIterator(WellFormednessArg);
 
 let check_well_formedness = ({statements}) => {
-  List.iter(WellFormednessIterator.iter_toplevel_stmt, statements);
+  ();
+    // List.iter(WellFormednessIterator.iter_toplevel_stmt, statements);
 };
```
Comment (jozanza) [inline compiler/src/typed/typed_well_formedness.re:159]: Great function 10/10

**Review by ospencer — APPROVED** [review]
Hooray! 🥳 🎉

**Review by marcusroberts — APPROVED** [review]
I'm not going to spot bugs, but nice work!

**Review by peblair — CHANGES_REQUESTED** [review]
I think this is mostly good; just a couple of small comments. Also, as discussed, could you or @ospencer update `memory_management.md` (or whichever file is most appropriate) to document the semantics of `@unsafe`?

File: compiler/src/codegen/compcore.re [hunk 2]
```diff
@@ -986,29 +1078,33 @@ let tee_swap =
...
-      Concatlist.list_of_t(
-        singleton(set_swap(wasm_mod, env, 0, arg, ~skip_incref=false))
-        @ cleanup_local_slot_instructions(wasm_mod, env)
-        +@ [get_swap(wasm_mod, env, 0)],
-      );
-    };
-  ret;
+  switch (rtype) {
+  | [_] when Config.no_gc^ => arg
+  | [Types.StackAllocated(_)] => arg
```
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:1098]: That's actually a bug, and an annoying one too 😢 

I'll have it patched in a sec

**Review by peblair — APPROVED** [review]
:shipit:

File: compiler/src/middle_end/anf_utils.rei [hunk 3]
```diff
@@ -5,7 +5,7 @@ let anf_free_vars: anf_expression => Ident.Set.t;
 let comp_free_vars: comp_expression => Ident.Set.t;
 let imm_free_vars: imm_expression => Ident.Set.t;
 
-let anf_count_vars: anf_expression => (int, int, int, int);
-let comp_count_vars: comp_expression => (int, int, int, int);
+let anf_count_vars: anf_expression => (int, int, int, int, int);
```
[the remark this answers is not in the document]
Comment (phated) [reply compiler/src/middle_end/anf_utils.rei:8]: I'd love that cleanup in the future.  @ospencer can you add an issue?

**Review by phated — APPROVED** [review]
Amazing. Primarily reviewed the code and then skimmed a couple snapshots. Top work @ospencer and @jozanza

File: compiler/src/middle_end/anf_utils.rei (hunk 3 above)
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/middle_end/anf_utils.rei:8]: Got you in #1155

**Pull request merged**